### PR TITLE
fix InsertOrdMap

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Fix the broken InsertOrdMap

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala [hunk 1]
```diff
@@ -45,6 +49,6 @@ object InsertOrdMap {
   def empty[K, V]: InsertOrdMap[K, V] = Empty.asInstanceOf[InsertOrdMap[K, V]]
 
   def apply[K, V](entries: (K, V)*): InsertOrdMap[K, V] =
-    new InsertOrdMap(entries.map(_._1)(breakOut), HashMap(entries: _*))
+    (empty[K, V] /: entries)(_ + _)
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala]: Here is the actual fix.

File: daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/InsertOrdMapTest.scala [hunk 2]
```diff
@@ -18,5 +18,8 @@ class InsertOrdMapTest extends WordSpec with Matchers {
       InsertOrdMap(1 -> "a", 2 -> "b").toSeq shouldEqual Seq(1 -> "a", 2 -> "b")
       InsertOrdMap(2 -> "b", 1 -> "a").toSeq shouldEqual Seq(2 -> "b", 1 -> "a")
     }
+    "drop duplicate" in {
+      InsertOrdMap(1 -> "a", 2 -> "b", 1 -> "c").toSeq shouldEqual Seq(1 -> "c", 2 -> "b")
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/test/scala/com/digitalasset/daml/lf/data/InsertOrdMapTest.scala]: Here is the test for the fix

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala [hunk 3]
```diff
@@ -45,6 +49,6 @@ object InsertOrdMap {
   def empty[K, V]: InsertOrdMap[K, V] = Empty.asInstanceOf[InsertOrdMap[K, V]]
 
   def apply[K, V](entries: (K, V)*): InsertOrdMap[K, V] =
-    new InsertOrdMap(entries.map(_._1)(breakOut), HashMap(entries: _*))
+    entries.foldLeft(empty[K, V])(_ + _)
```
Comment (remyhaemmerle-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala]: This is the fix

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala (hunk 1 above)
Comment (SamirTalwar) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala]: `/:` is deprecated in Scala 2.12.10 and 2.13; let's use `foldLeft` instead. 🙂

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala [hunk 4]
```diff
@@ -15,25 +14,30 @@ import scala.collection.immutable.{HashMap, Map, Queue}
...
-      new InsertOrdMap(keys, hashMap + kv)
+  override def updated[V1 >: V](key: K, value: V1): InsertOrdMap[K, V1] =
+    if (hashMap.contains(key))
+      new InsertOrdMap(keys, hashMap.updated(key, value))
     else
-      new InsertOrdMap(keys :+ kv._1, hashMap + kv)
+      new InsertOrdMap(keys :+ key, hashMap.updated(key, value))
 
-  override def -(k: Key): InsertOrdMap[Key, Value] =
+  override def +[V >: V](kv: (K, V)): InsertOrdMap[K, V] = {
```
Comment (SamirTalwar) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala]: Think you want `V1` here.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala]: fixed

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/InsertOrdMap.scala]: fixed

**Pull request merged**